fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/keyspace/KeySpaceDirectoryTest.java

```diff
@@ -547,6 +548,123 @@ public void testSeesMetadataUpdates() {
+                    .build());
+            RecordCursor<KeySpacePath> cursor = RecordCursor.flatMapPipelined(
+                    outerContinuation ->
+                            root.path("root").listAsync(context, "a", outerContinuation, props)
+                                    .map(value -> {
+                                        sleep(1L);
+                                        return value;
+                                    }),
+                    (aPath, innerContinuation) ->
+                            aPath.add("b", 0).listAsync(context, "c", innerContinuation, props),
+                    null,
+                    10
+            );
+
+            int count = 0;
+            while (cursor.hasNext()) {
+                cursor.next();
+                ++count;
+            }
+
+            assertEquals(RecordCursor.NoNextReason.TIME_LIMIT_REACHED, cursor.getNoNextReason());
+            // With a 1ms delay we should read no more than 5 "a" values (there are a total of 10)
+            // and each "c" value has 4 values. so we shouldn't have been able to read more than 40
+            // total values.
+            assertTrue(count <= 40, "Read too many values, query should have timed out");
+        }
+    }
+
+    @Test
+    public void testListObeysReturnedRowAndScanLimits() {
+        KeySpace root = new KeySpace(
+                new KeySpaceDirectory("root", KeyType.STRING, "root-" + random.nextInt(Integer.MAX_VALUE))
+                        .addSubdirectory(new KeySpaceDirectory("a", KeyType.LONG)
+                                .addSubdirectory(new KeySpaceDirectory("b", KeyType.LONG))));
+
+        final FDBDatabase database = FDBDatabaseFactory.instance().getDatabase();
+        try (FDBRecordContext context = database.openContext()) {
+            Transaction tr = context.ensureActive();
+            for (int i = 0; i < 5; i++) {
+                for (int j = 0; j < 2; j++) {
+                    tr.set(root.path("root")
+                            .add("a", i)
+                            .add("b", j)
+                            .toTuple(context).pack(), Tuple.from(i + j).pack());
+                }
+            }
+            tr.commit().join();
+        }
+
+        doLimitedScan(database, root, 5, Integer.MAX_VALUE, RecordCursor.NoNextReason.RETURN_LIMIT_REACHED);
+        doLimitedScan(database, root, Integer.MAX_VALUE, 5, RecordCursor.NoNextReason.SCAN_LIMIT_REACHED);
+    }
+
+    private void doLimitedScan(FDBDatabase database, KeySpace root, int returnedRowLimit, int scannedRecordLimit,
+                               RecordCursor.NoNextReason noNextReason) {
+        try (FDBRecordContext context = database.openContext()) {
+            // Iteration will inject a 1ms pause in each "a" value we iterate over (there are 10 of them)
+            // so we want to make the time limit long enough to make *some* progress, but short enough to
+            // to make sure we cannot get them all.
```

Comment:
I don't think that this comment applies here (I don't see a time limit). Perhaps it was mistakenly copied from above?

Reply:
Indeed 'twas a copy-n-paste error.